Lib/fontmake/font_project.py

```diff
@@ -669,50 +670,93 @@ def interpolate_instance_ufos(
-                "attribute"
-            )
+        logger.info(
+            "Interpolating master UFOs from designspace and applying instance data"
+        )
+        designspace = designspaceLib.DesignSpaceDocument.fromfile(designspace.path)
+        generator = fontmake.instantiator.Instantiator.from_designspace(
+            designspace, round_geometry=round_instances
+        )
+        if expand_features_to_instances:
+            feature_text = parseLayoutFeatures(designspace.default.font).asFea()
 
-        with temporarily_disabling_axis_maps(designspace.path) as temp_designspace_path:
-            builder = DesignSpaceDocumentReader(
-                temp_designspace_path,
-                ufoVersion=3,
-                roundGeometry=round_instances,
-                verbose=True,
-            )
-            logger.info("Interpolating master UFOs from designspace")
-            if include is not None:
-                instances = self._search_instances(designspace, pattern=include)
-                for instance_name in instances:
-                    builder.readInstance(("name", instance_name))
-                filenames = set(instances.values())
-            else:
-                builder.readInstances()
-                filenames = None  # will include all instances
+        for instance in designspace.instances:
+            if not instance.lib.get("com.schriftgestaltung.export", True):
+                continue
 
-        logger.info("Applying instance data from designspace")
-        instance_ufos = apply_instance_data(designspace, include_filenames=filenames)
+            if include is not None and not fullmatch(include, instance.name):
+                continue
 
-        if expand_features_to_instances:
-            logger.debug("Expanding features to instance UFOs")
-            master_source = next(
-                (s for s in designspace.sources if s.copyFeatures), None
-            )
-            if not master_source:
-                raise ValueError("No source is designated as the master for features.")
-            else:
-                master_source_font = builder.sources[master_source.name][0]
-                master_source_features = parseLayoutFeatures(master_source_font).asFea()
-                for instance_ufo in instance_ufos:
-                    instance_ufo.features.text = master_source_features
-                    instance_ufo.save()
+            instance.font = generator.generate_instance(instance)
+
+            apply_instance_data_to_ufo(instance.font, instance, designspace)
 
-        return instance_ufos
+            if expand_features_to_instances:
+                logger.debug("Expanding features to instance UFO %s", instance.filename)
+                instance.font.features.text = feature_text
+
+            assert (
+                instance.filename is not None
+            ), "FIXME: Make fontmake not rely on instance filenames"
```

Comment:
what do you mean "not rely on instance filenames"? how does fontmake know where to save the instances if requested to generate them?

Reply:
My concern is what happens if `filename` is not specified.